File: lib/data/exams/ae-2.dart
import '../../models/ihk_exam_model.dart';

final ae2Exam = IHKExam(
  id: 'ae-2',
  title: 'AE Prüfung 2 - Sommer 2017',
  year: 2017,
  season: 'Sommer',
  duration: 90,
  totalPoints: 100,
  company: 'SecureID GmbH',
  scenario:
      '''Sie sind Mitarbeiter/-in der SecureID GmbH, Darmstadt, einem Softwaredienstleister im Bereich biometrische Sicherheitssysteme. Die SecureID GmbH erstellt Software zur Erfassung und Auswertung verschiedener biometrischer Daten.

Sie sollen vier der folgenden fünf Aufgaben erledigen:
1. Ein UML-Klassendiagramm erstellen
2. Eine Funktion zur Auswertung von Iris-Scans erstellen
3. Ein UML-Aktivitätsdiagramm erstellen
4. Ein relationales Datenmodell erstellen
5. SQL-Anweisungen für eine Datenbank erstellen''',
  sections: [
    ExamSection(
      id: 'hs1',
      title: 'Handlungsschritt 1: UML-Klassendiagramm (25 Punkte)',
      totalPoints: 25,
      questions: [
        ExamQuestion(
          id: 'hs1-q1',
          title: 'UML-Klassendiagramm - Biometrische Scans',
          description:
              '''Die SecureID GmbH soll eine Software zur Erkennung und Speicherung von Iris-Scans und Netzhaut-Scans erstellen.

Für eine Person sollen vom linken und rechten Auge jeweils folgende Scans gespeichert werden:
- Iris-Scan
- Netzhaut-Scan

Zu jedem Scan sollen ein Bild und ein String gespeichert werden.

Die Zeichenkette enthält Beschreibungen derjenigen Merkmale des Scans, die beim Vergleich von Scans verwendet werden.

Die Zeichenkette wird von der Methode berechneMerkmal() anhand des Bildes berechnet.

Die Algorithmen zur Berechnung der Zeichenketten sind für Iris-Scan und Netzhaut-Scan unterschiedlich.

Es existiert bereits folgende Klasse Scan:

┌─────────────────────────────┐
│ Scan                        │
├─────────────────────────────┤
│ - bild: Bild                │
│ - merkmal: String           │
├─────────────────────────────┤
│ + berechneMerkmal()         │
└─────────────────────────────┘

AUFGABE:
Erstellen Sie ein vollständiges UML-Klassendiagramm mit:
- Klasse Person
- Klasse Scan (gegeben)
- Klassen IrisScan und NetzhautScan (erben von Scan)
- Alle notwendigen Beziehungen und Kardinalitäten''',
          type: QuestionType.diagram,
          points: 25,
          hint:
              'Nutzen Sie Vererbung für IrisScan und NetzhautScan. Eine Person hat 4 Scans (2 pro Auge).',
        ),
      ],
    ),
    ExamSection(
      id: 'hs2',
      title: 'Handlungsschritt 2: Algorithmus Iris-Scan (25 Punkte)',
      totalPoints: 25,
      questions: [
        ExamQuestion(
          id: 'hs2-q1',
          title: 'Algorithmus - Iris-Scan Auswertung',
          description:
              '''Die SecureID GmbH benötigt einen Algorithmus zur Auswertung von Iris-Scans.

Ein Iris-Scan wird als Array von Grauwerten gespeichert (0-255).

AUFGABE:
Entwickeln Sie einen Algorithmus (Pseudocode oder Struktogramm), der:

1. Das Array durchläuft
2. Die Anzahl der hellen Pixel (Wert > 200) zählt
3. Die Anzahl der dunklen Pixel (Wert < 50) zählt
4. Das Verhältnis hell/dunkel berechnet
5. Wenn Verhältnis > 2.0: "Guter Scan"
6. Wenn Verhältnis < 0.5: "Schlechter Scan"
7. Sonst: "Normaler Scan"

BEISPIEL:
Array: [230, 45, 210, 30, 240, 20]
Hell (>200): 3 Pixel
Dunkel (<50): 3 Pixel
Verhältnis: 3/3 = 1.0
Ergebnis: "Normaler Scan"''',
          type: QuestionType.code,
          points: 25,
          hint: 'Nutzen Sie Zählvariablen und eine Schleife.',
        ),
      ],
    ),
    ExamSection(
      id: 'hs3',
      title: 'Handlungsschritt 3: UML-Aktivitätsdiagramm (25 Punkte)',
      totalPoints: 25,
      questions: [
        ExamQuestion(
          id: 'hs3-q1',
          title: 'UML-Aktivitätsdiagramm - Authentifizierung',
          description:
              '''Modellieren Sie den Authentifizierungsprozess mit biometrischen Daten.

ABLAUF:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
1. Benutzer startet Login
2. System fordert biometrischen Scan an
3. Benutzer wählt: Iris-Scan ODER Fingerabdruck
4. Bei Iris-Scan:
   - Kamera aktivieren
   - Scan durchführen
   - Qualität prüfen (wenn schlecht → zurück zu Schritt 2)
5. Bei Fingerabdruck:
   - Scanner aktivieren
   - Scan durchführen
6. System vergleicht mit Datenbank
7. Bei Übereinstimmung > 95%: Zugang gewähren
8. Sonst: Zugang verweigern
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

AUFGABE:
Erstellen Sie ein UML-Aktivitätsdiagramm mit:
- Start/Endknoten
- Entscheidungen
- Aktivitäten
- Schleifen (für Qualitätsprüfung)''',
          type: QuestionType.diagram,
          points: 25,
          hint:
              'Nutzen Sie eine Raute für die Scan-Auswahl und eine weitere für die Qualitätsprüfung.',
        ),
      ],
    ),
    ExamSection(
      id: 'hs4',
      title: 'Handlungsschritt 4: Datenmodell (25 Punkte)',
      totalPoints: 25,
      questions: [
        ExamQuestion(
          id: 'hs4-q1',
          title: 'ER-Diagramm - Zutrittskontrolle',
          description:
              '''Die SecureID GmbH möchte ein System zur Zutrittskontrolle entwickeln.

ANFORDERUNGEN:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
ENTITÄTEN:
- Mitarbeiter (MA_ID, Name, Abteilung)
- Raum (Raum_ID, Bezeichnung, Gebäude, Sicherheitsstufe)
- Zutritt (Zutritt_ID, Zeitpunkt, Erfolg)
- Zutrittsberechtigung (von_Datum, bis_Datum)

BEZIEHUNGEN:
- Ein Mitarbeiter kann viele Zutrittsversuche machen
- Ein Raum hat viele Zutrittsversuche
- Ein Mitarbeiter hat Berechtigungen für mehrere Räume
- Ein Raum kann von mehreren Mitarbeitern betreten werden
- Eine Berechtigung ist zeitlich befristet (von-bis)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

AUFGABE:
Erstellen Sie ein ER-Diagramm mit allen Entitäten, Beziehungen und Kardinalitäten.''',
          type: QuestionType.diagram,
          points: 25,
          hint:
              'Die Berechtigung ist eine Beziehungsentität zwischen Mitarbeiter und Raum.',
        ),
      ],
    ),
    ExamSection(
      id: 'hs5',
      title: 'Handlungsschritt 5: SQL-Abfragen (25 Punkte)',
      totalPoints: 25,
      questions: [
        ExamQuestion(
          id: 'hs5-intro',
          title: 'Datenbankschema - SecureID System',
          description: '''TABELLE: Mitarbeiter
| MA_ID | Name      | Abteilung |
|-------|-----------|-----------|
| 1     | Müller    | IT        |
| 2     | Schmidt   | HR        |
| 3     | Weber     | IT        |

TABELLE: Raum
| Raum_ID | Bezeichnung | Sicherheitsstufe |
|---------|-------------|------------------|
| 101     | Server      | 3                |
| 102     | Büro        | 1                |
| 103     | Labor       | 2                |

TABELLE: Zutritt
| Zutritt_ID | MA_ID | Raum_ID | Zeitpunkt           | Erfolg |
|------------|-------|---------|---------------------|--------|
| 1          | 1     | 101     | 2024-01-20 08:00:00 | true   |
| 2          | 2     | 102     | 2024-01-20 09:00:00 | true   |
| 3          | 1     | 103     | 2024-01-20 10:00:00 | false  |''',
          type: QuestionType.info,
          points: 0,
        ),
        ExamQuestion(
          id: 'hs5-a',
          title: 'Aufgabe a) Erfolgreiche Zutritte pro Mitarbeiter (8 Punkte)',
          description:
              '''Erstellen Sie eine SQL-Abfrage, die für jeden Mitarbeiter die Anzahl der erfolgreichen Zutritte ausgibt.

ERWARTETE AUSGABE:
| Name    | AnzahlZutritte |
|---------|----------------|
| Müller  | 1              |
| Schmidt | 1              |
| Weber   | 0              |''',
          type: QuestionType.code,
          points: 8,
          hint: 'Nutzen Sie COUNT() und GROUP BY. Denken Sie an LEFT JOIN.',
        ),
        ExamQuestion(
          id: 'hs5-b',
          title: 'Aufgabe b) Räume mit höchster Sicherheitsstufe (5 Punkte)',
          description:
              '''Geben Sie alle Räume aus, die die höchste Sicherheitsstufe haben.''',
          type: QuestionType.code,
          points: 5,
          hint: 'Nutzen Sie eine Unterabfrage mit MAX().',
        ),
        ExamQuestion(
          id: 'hs5-c',
          title:
              'Aufgabe c) Fehlgeschlagene Zutritte der IT-Abteilung (7 Punkte)',
          description:
              '''Listen Sie alle fehlgeschlagenen Zutrittsversuche von Mitarbeitern der IT-Abteilung auf.

Sortiert nach Zeitpunkt (neueste zuerst).''',
          type: QuestionType.code,
          points: 7,
          hint:
              'JOIN Mitarbeiter und Zutritt, filtern nach Abteilung und Erfolg.',
        ),
        ExamQuestion(
          id: 'hs5-d',
          title:
              'Aufgabe d) Durchschnittliche Sicherheitsstufe pro Abteilung (5 Punkte)',
          description:
              '''Berechnen Sie die durchschnittliche Sicherheitsstufe der Räume, die von jeder Abteilung betreten wurden (nur erfolgreiche Zutritte).''',
          type: QuestionType.code,
          points: 5,
          hint: 'Nutzen Sie AVG() mit mehreren JOINs.',
        ),
      ],
    ),
  ],
);

File: web/data/exams/ae-2.ts
import { Exam } from "../exam-types";

export const ae2: Exam = {
    id: "ae-2",
    title: "AE Prüfung 2 - Sommer 2017",
    year: 2017,
    season: "Sommer",
    company: "SecureID GmbH",
    duration: 90,
    totalPoints: 100,
    level: "ap2",
    fachrichtung: "ae",
    difficulty: "mittel",
    tags: ["uml-klasse", "uml-aktivität", "normalisierung", "sql", "algorithmen"],
    sectionsToChoose: 4,
    scenario: `Sie sind Mitarbeiter/-in der SecureID GmbH, Darmstadt, einem Softwaredienstleister im Bereich biometrische Sicherheitssysteme. Die SecureID GmbH erstellt Software zur Erfassung und Auswertung verschiedener biometrischer Daten.

Sie sollen vier der folgenden fünf Aufgaben erledigen:
1. Ein UML-Klassendiagramm erstellen
2. Eine Funktion zur Auswertung von Iris-Scans erstellen
3. Ein UML-Aktivitätsdiagramm erstellen
4. Ein relationales Datenmodell erstellen
5. SQL-Anweisungen für eine Datenbank erstellen`,
    sections: [
        {
            id: "hs1",
            title: "Handlungsschritt 1: UML-Klassendiagramm (25 Punkte)",
            totalPoints: 25,
            questions: [
                {
                    id: "hs1-aa",
                    title: "Aufgabe a-aa) Assoziation beschreiben (2 Punkte)",
                    description: `Die SecureID GmbH soll eine Software zur Erkennung und Speicherung von Iris-Scans und Netzhaut-Scans erstellen. Zur Vorbereitung der Programmierung soll ein UML-Klassendiagramm erstellt werden.

Für eine Person sollen vom linken und rechten Auge jeweils folgende Scans gespeichert werden:

I1: Iris-Scan
N1: Netzhaut-Scan

Zu jedem Scan sollen ein Bild und ein String gespeichert werden.

Die Zeichenkette enthält Beschreibungen derjenigen Merkmale des Scans, die beim Vergleich von Scans verwendet werden.

Die Zeichenkette wird von der Methode berechneMerkmal() anhand des Bildes berechnet.

Die Algorithmen zur Berechnung der Zeichenketten sind für Iris-Scan und Netzhaut-Scan unterschiedlich.

Es existiert bereits folgende Klasse Scan, die für das Klassendiagramm verwendet werden soll:

┌─────────────────────────────┐
│ Scan                        │
├─────────────────────────────┤
│ - bild: Bild                │
│ - merkmal: String           │
├─────────────────────────────┤
│ + berechneMerkmal()         │
└─────────────────────────────┘

In einem UML-Klassendiagramm können die folgenden Beziehungen vorkommen.
Beschreiben Sie jeweils kurz:

aa) Assoziation`,
                    type: "freeText",
                    points: 2,
                    hint: `Assoziation: Eine allgemeine Beziehung zwischen zwei Klassen. Objekte der einen Klasse kennen Objekte der anderen Klasse.`,
                    tags: ["uml-klasse", "theorie", "assoziation"],
                },
                {
                    id: "hs1-ab",
                    title: "Aufgabe a-ab) Vererbung beschreiben (2 Punkte)",
                    description: `ab) Vererbung`,
                    type: "freeText",
                    points: 2,
                    hint: `Vererbung: Eine Klasse übernimmt Attribute und Methoden von einer anderen Klasse. "ist-ein"-Beziehung.`,
                    tags: ["uml-klasse", "theorie", "vererbung", "oop"],
                },
                {
                    id: "hs1-ac",
                    title: "Aufgabe a-ac) Komposition beschreiben (2 Punkte)",
                    description: `ac) Komposition`,
                    type: "freeText",
                    points: 2,
                    hint: `Komposition: Starke "Teil-von"-Beziehung. Das Teil kann ohne das Ganze nicht existieren.`,
                    tags: ["uml-klasse", "theorie", "komposition"],
                },
                {
                    id: "hs1-b",
                    title: "Aufgabe b) UML-Klassendiagramm erstellen (19 Punkte)",
                    description: `Erstellen Sie ein UML-Klassendiagramm, das ...

- die Klassen Person, Auge, IrisBereich, NetzhautBereich, Scan, IrisScan, NetzhautScan darstellt.
- die Beziehungen zwischen den Klassen mit ihren Kardinalitäten angibt.
- Geben Sie an, in welchen Klassen die Methode berechneMerkmal() überschrieben werden muss.`,
                    type: "diagram",
                    points: 19,
                    hint: `Struktur:
- Person hat 2 Augen (Komposition)
- Auge hat 1 IrisBereich und 1 NetzhautBereich (Komposition)
- IrisScan und NetzhautScan erben von Scan
- berechneMerkmal() muss in IrisScan und NetzhautScan überschrieben werden`,
                    diagram: { mode: "uml-class" },
                    tags: ["uml-klasse", "vererbung", "komposition", "kardinalitäten"],
                },
            ],
        },
        {
            id: "hs2",
            title: "Handlungsschritt 2: Algorithmus Iris-Auswertung (25 Punkte)",
            totalPoints: 25,
            questions: [
                {
                    id: "hs2-q1",
                    title: "Algorithmus - Prozedur auswertung",
                    description: `Um herauszufinden, von welcher Person ein Iris-Scan stammt, soll dieser mit Iris-Scans in einer Datenbank verglichen werden.

Die vorhandene Funktion suche(scan) gibt ein Array treffer aus, das zu jedem gefundenen Iris-Scan einen Score, eine Personen-ID und eine Auge-ID enthält.

PARAMETER DER PROZEDUR:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
| Parameter | Beschreibung                                        |
|-----------|-----------------------------------------------------|
| scan      | Zeichenkette; Werte des Iris-Scan-Bildes            |
| schwelle  | ganzzahliger Wert; 1 bis 100; Score-Schwellenwert   |
| auge      | 0 = Unbekannt; 1 = rechtes Auge; 2 = linkes Auge    |
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

VERFÜGBARE FUNKTIONEN:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
| Funktion           | Beschreibung                                 |
|--------------------|----------------------------------------------|
| suche(scan)        | Gibt Array vom Typ Treffer zurück            |
| laenge(array)      | Liefert die Länge des Arrays                 |
| loesche(array,pos) | Löscht Element an Position pos               |
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

BEISPIEL:

Array treffer von suche(scan):          Rückgabe bei schwelle=80, auge=1:
┌───────┬──────────┬────────┐           ┌───────┬──────────┬────────┐
│ score │ idPerson │ idAuge │           │ score │ idPerson │ idAuge │
├───────┼──────────┼────────┤           ├───────┼──────────┼────────┤
│ 85    │ 93334    │ 1      │           │ 98    │ 56446    │ 1      │
│ 80    │ 48774    │ 2      │           │ 85    │ 93334    │ 1      │
│ 98    │ 56446    │ 1      │           │ 81    │ 49982    │ 1      │
│ 71    │ 33961    │ 2      │           └───────┴──────────┴────────┘
│ 21    │ 73447    │ 1      │
│ 81    │ 49982    │ 1      │
└───────┴──────────┴────────┘

AUFGABE:
    auswertung(scan: String, schwelle: Integer, auge: Integer) : Treffer[]`,
                    type: "code",
                    points: 25,
                    hint: `1. treffer = suche(scan)
2. Filterphase (von hinten nach vorne):
   - score <= schwelle → löschen
   - auge != 0 UND idAuge != auge → löschen
3. Bubble Sort nach score ABSTEIGEND
4. Rückgabe: treffer`,
                    tags: ["algorithmen", "pseudocode", "bubble-sort", "filterung"],
                },
            ],
        },
        {
            id: "hs3",
            title: "Handlungsschritt 3: UML-Aktivitätsdiagramm & Minimum (25 Punkte)",
            totalPoints: 25,
            questions: [
                {
                    id: "hs3-a",
                    title: "Aufgabe a) UML-Aktivitätsdiagramm (20 Punkte)",
                    description: `Die SecureID GmbH soll ein System zur Iris-Scan-Recherche erstellen.

Die Recherche soll wie folgt organisiert werden:

- Ein Auftraggeber schickt einen Iris-Scan (IS) zur Identifizierung an den Operator.
- Der Operator prüft, ob die Qualität des IS in Ordnung ist.
- Ist die Qualität nicht ok → Info an Auftraggeber → Ende
- Ist die Qualität ok → Operator führt Suche durch
- Keine Treffer → Info an Auftraggeber → Ende
- Treffer gefunden → PARALLEL: Report erstellen UND Daten an Supervisor
- Supervisor protokolliert und schickt Info zurück
- Nach Report UND Info: Operator sendet Report an Auftraggeber → Ende

Erstellen Sie ein UML-Aktivitätsdiagramm mit Schwimmbahnen.`,
                    type: "diagram",
                    points: 20,
                    hint: `Schwimmbahnen: Auftraggeber | Operator | Supervisor

Wichtige Elemente:
- Fork für Parallelisierung
- Join für Synchronisation
- Zwei Entscheidungen (Qualität, Treffer)`,
                    diagram: { mode: "uml-activity" },
                    tags: ["uml-aktivität", "fork-join", "swimlanes"],
                },
                {
                    id: "hs3-b",
                    title: "Aufgabe b) Minimum ermitteln (5 Punkte)",
                    description: `Array treffer:
┌───────┬──────────┐
│ score │ idPerson │
├───────┼──────────┤
│ 21    │ 73447    │
│ 85    │ 93334    │
│ 80    │ 48774    │
│ 98    │ 56446    │
│ 81    │ 49982    │
└───────┴──────────┘

Ermitteln Sie den minimalen Score-Wert.

Beispiel-Ausgabe: minimaler Score = 21`,
                    type: "code",
                    points: 5,
                    hint: "Initialisieren Sie minimum mit dem ersten Element und durchlaufen Sie dann das Array.",
                    tags: ["algorithmen", "pseudocode", "minimum", "schleife"],
                },
            ],
        },
        {
            id: "hs4",
            title: "Handlungsschritt 4: Relationales Datenmodell (25 Punkte)",
            totalPoints: 25,
            questions: [
                {
                    id: "hs4-q1",
                    title: "Relationales Datenmodell in 3. Normalform",
                    description: `Die SecureID GmbH soll für eine Sicherheitsfirma eine Datenbank erstellen.

Erfassung von Vorfällen (Excel-Tabelle):
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
| Vorfall | MA-ID | Anrede | MA-Name     | Geb.datum  | Adresse          |
|---------|-------|--------|-------------|------------|------------------|
| 501     | 7823  | Herr   | Schmidt, T. | 15.03.1985 | 60325 Frankfurt  |
| 502     | 4521  | Herr   | Müller, K.  | 22.07.1990 | 60489 Frankfurt  |
| 503     | 7823  | Herr   | Schmidt, T. | 15.03.1985 | 60325 Frankfurt  |
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

| Vorfall-Typ    | Datum      | Dokument              | Bearbeiter    |
|----------------|------------|-----------------------|---------------|
| Einbruch       | 12.04.2024 | Personalausweis, FS   | Hansen, Klaus |
| Alarmauslösung | 18.04.2024 | Personalausweis       | Müller, Marcel|
| Sabotage       | 25.04.2024 | Reisepass, FS         | Hansen, Klaus |
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Erstellen Sie ein relationales Datenmodell in der 3. Normalform.

- Selbsterklärende Namen für Tabellen und Attribute
- PK und FK kennzeichnen
- Beziehungen mit Kardinalitäten

Hinweis: Adresse muss nicht normalisiert werden.`,
                    type: "diagram",
                    points: 25,
                    hint: "Welche Daten wiederholen sich? Dokumente kommen mehrfach vor - wie lösen Sie das auf?",
                    tags: ["normalisierung", "3nf", "datenmodell"],
                },
            ],
        },
        {
            id: "hs5",
            title: "Handlungsschritt 5: SQL-Abfragen (25 Punkte)",
            totalPoints: 25,
            questions: [
                {
                    id: "hs5-intro",
                    title: "Datenbankschema - Zugangskontrolle",
                    description: `Datenbank für Zugangskontrolle mit Testdaten:

Person:
| PersID | Nachname | Vorname | Strasse      | Plz   | Ort      |
|--------|----------|---------|--------------|-------|----------|
| 101    | Müller   | Max     | Müllerweg 1  | 52335 | Köln     |
| 202    | Meier    | Willi   | Testweg 12   | 43333 | Dortmund |
| 404    | Wester   | Klaus   | Hauptstr. 13 | 55667 | Köln     |

Zugang:
| RaumID | PersID | ZeitVon | ZeitBis |
|--------|--------|---------|---------|
| 1      | 101    | 08:00   | 10:00   |
| 1      | 202    | 10:00   | 14:00   |
| 2      | 101    | 14:00   | 18:00   |
| 5      | 202    | 08:00   | 18:00   |

Raum:
| RaumID | RaumTyp          | GebID | MerkID |
|--------|------------------|-------|--------|
| 1      | Besprechungsraum | 2     | 1      |
| 2      | Labor            | 2     | 2      |
| 5      | Besprechungsraum | 1     | 1      |

Gebaeude:
| GebID | Bezeichnung | Strasse          | Plz   | Ort  |
|-------|-------------|------------------|-------|------|
| 1     | Forschung H | Heinrich-Str. 12 | 50501 | Köln |
| 2     | Forschung U | Heinrich-Str. 14 | 50501 | Köln |

Merkmal:
| MerkID | Merkmal       |
|--------|---------------|
| 1      | Iris-Scan     |
| 2      | Fingerabdruck |`,
                    type: "info",
                    points: 0,
                    hint: "Datenbankschema für die folgenden SQL-Aufgaben.",
                },
                {
                    id: "hs5-a",
                    title: "Aufgabe a) Gebäude mit Räumen (5 Punkte)",
                    description: `Liste aller Gebäude mit deren Räumen, sortiert nach Bezeichnung und RaumTyp.`,
                    type: "code",
                    points: 5,
                    hint: "Welche Tabellen müssen Sie JOINen? Wie sortieren Sie nach mehreren Spalten?",
                    tags: ["sql", "join", "order-by"],
                },
                {
                    id: "hs5-b",
                    title: "Aufgabe b) RIGHT JOIN Zugang (5 Punkte)",
                    description: `Liste aller Zugangsdaten mit dazugehörigen Personendaten.`,
                    type: "code",
                    points: 5,
                    hint: "Was ist der Unterschied zwischen LEFT JOIN und RIGHT JOIN?",
                    tags: ["sql", "right-join", "join"],
                },
                {
                    id: "hs5-c",
                    title: "Aufgabe c) Anzahl Räume pro Merkmal (6 Punkte)",
                    description: `Anzahl der Räume pro Zugangskontroll-Merkmal.

Erwartete Ausgabe:
| Merkmal       | AnzahlRaeume |
|---------------|--------------|
| Iris-Scan     | 2            |
| Fingerabdruck | 6            |`,
                    type: "code",
                    points: 6,
                    hint: "Welche Aggregatfunktion zählt Zeilen? Wie gruppieren Sie nach Merkmal?",
                    tags: ["sql", "count", "group-by", "aggregation"],
                },
                {
                    id: "hs5-d",
                    title: "Aufgabe d) Zugangsdaten Max Müller (6 Punkte)",
                    description: `Liste der Zugangsdaten von Max Müller (nur Name bekannt).`,
                    type: "code",
                    points: 6,
                    hint: "Wie filtern Sie nach einem bestimmten Namen? Welche Tabellen verknüpfen Person und Zugang?",
                    tags: ["sql", "where", "join"],
                },
                {
                    id: "hs5-e",
                    title: "Aufgabe e) PLZ-Bereich 50000-59999 (3 Punkte)",
                    description: `Liste aller Personen aus PLZ-Gebiet 50000-59999.`,
                    type: "code",
                    points: 3,
                    hint: "BETWEEN eignet sich gut für Bereichsabfragen.",
                    tags: ["sql", "between", "where"],
                },
            ],
        },
    ],
};